httpcore/_async/http11.py

```diff
@@ -344,15 +353,16 @@ async def __aiter__(self) -> AsyncIterator[bytes]:
             # If we get an exception while streaming the response,
             # we want to close the response (and possibly the connection)
             # before raising that exception.
-            with AsyncShieldCancellation():
-                await self.aclose()
+            if self._connection._connection_should_close():
+                await self._connection.aclose()
             raise exc
 
     async def aclose(self) -> None:
-        if not self._closed:
-            self._closed = True
-            async with Trace("response_closed", logger, self._request):
-                await self._connection._response_closed()
+        async with Trace("response_closed", logger, self._request, kwargs={}):
+            if not self._closed:
+                self._closed = True
+                if self._connection._connection_should_close():
```

Comment:
Can the await inside the `Trace.__aenter__` cause issues here? Ie can there be a cancellation with the await inside it that would cause a skip of the `self._closed`-checks and `self._connection._connection_should_close()`?

Reply:
yes this is a problem, It's also a problem if the trace function raises, I think you'll need to do this:

```python
# synchronization.py
aclose_forcefully = anyio.aclose_forcefully
def close_forcefully(sock):
    sock.close()
```

```python
    async def aclose(self):
        entered_cmgr = False
        try:
            async with Trace("response_closed", logger, self._request, kwargs={}):
                entered_cmgr = True
                if not self._closed:
                    self._closed = True
                    if self._connection._connection_should_close():
                        await self._connection.aclose()
        except BaseException:
            if entered_cmgr:
                raise
            if not self._closed:
                self._closed = True
                if self._connection._connection_should_close():
                    await aclose_forcefully(self._connection)
            raise

```